File: CS/CalDAVServer.SqlStorage.AspNet/AuthenticationModuleBase.cs
using System;
using System.Security;
using System.Security.Principal;
using System.Text;
using System.Web;
using System.Web.Security;

namespace CalDAVServer.SqlStorage.AspNet
{
    /// <summary>
    /// Base class for challenge/response authentication ASP.NET modules, like Digest, Basic.
    /// </summary>
    public abstract class AuthenticationModuleBase : IHttpModule
    {

        public void Init(HttpApplication application)
        {
            application.AuthenticateRequest += App_OnAuthenticateRequest;
            application.EndRequest += App_OnEndRequest;
        }

        public void Dispose()
        {
        }

        
        protected abstract IPrincipal AuthenticateRequest(HttpRequest request);
        
        protected abstract string GetChallenge();

        protected abstract bool IsAuthorizationPresent(HttpRequest request);
        
        private void App_OnAuthenticateRequest(object source, EventArgs eventArgs)
        {
            if (IsAuthorizationPresent(HttpContext.Current.Request))
            {
                IPrincipal principal = AuthenticateRequest(HttpContext.Current.Request);
                if (principal != null)
                { // authenticated succesfully
                    HttpContext.Current.User = principal;
                }
                else
                { // invalid credentials
                    unauthorized();
                }
            }
            else
            {
                // To support Miniredirector/Web Folders on XP and Server 2003 as well as 
                // Firefox CORS requests, OPTIONS must be processed without authorization.
                if (HttpContext.Current.Request.HttpMethod == "OPTIONS")
                    return;

                unauthorized();
            }
        }
        
        private void App_OnEndRequest(object source, EventArgs eventArgs)
        {

            HttpApplication app = (HttpApplication)source;
            if (app.Response.StatusCode == 401)
            { // show login dialog
                app.Response.AppendHeader("WWW-Authenticate", GetChallenge());
            }
        }

        private static void unauthorized()
        {
            HttpResponse response = HttpContext.Current.Response;
            response.StatusCode = 401;
            response.StatusDescription = "Unauthorized";
            response.Write("401 Unauthorized");
        }
    }

}

File: VB/CalDAVServer.SqlStorage.AspNet/AuthenticationModuleBase.vb
Imports System
Imports System.Security
Imports System.Security.Principal
Imports System.Text
Imports System.Web
Imports System.Web.Security

''' <summary>
''' Base class for challenge/response authentication ASP.NET modules, like Digest, Basic.
''' </summary>
Public MustInherit Class AuthenticationModuleBase
    Implements IHttpModule

    Public Sub Init(application As HttpApplication) Implements IHttpModule.Init
        AddHandler application.AuthenticateRequest, AddressOf App_OnAuthenticateRequest
        AddHandler application.EndRequest, AddressOf App_OnEndRequest
    End Sub

    Public Sub Dispose() Implements IHttpModule.Dispose
    End Sub

    Protected MustOverride Function AuthenticateRequest(request As HttpRequest) As IPrincipal

    Protected MustOverride Function GetChallenge() As String

    Protected MustOverride Function IsAuthorizationPresent(request As HttpRequest) As Boolean

    Private Sub App_OnAuthenticateRequest(source As Object, eventArgs As EventArgs)
        If IsAuthorizationPresent(HttpContext.Current.Request) Then
            Dim principal As IPrincipal = AuthenticateRequest(HttpContext.Current.Request)
            If principal IsNot Nothing Then
                HttpContext.Current.User = principal
            Else
                unauthorized()
            End If
        Else
            If HttpContext.Current.Request.HttpMethod = "OPTIONS" Then Return
            unauthorized()
        End If
    End Sub

    Private Sub App_OnEndRequest(source As Object, eventArgs As EventArgs)
        Dim app As HttpApplication = CType(source, HttpApplication)
        If app.Response.StatusCode = 401 Then
            app.Response.AppendHeader("WWW-Authenticate", GetChallenge())
        End If
    End Sub

    Private Shared Sub unauthorized()
        Dim response As HttpResponse = HttpContext.Current.Response
        response.StatusCode = 401
        response.StatusDescription = "Unauthorized"
        response.Write("401 Unauthorized")
    End Sub
End Class
